oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSchemaReader.java

```diff
@@ -22,38 +22,110 @@
 import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Objects;
 import java.util.Set;
 
 /**
  * Oracle schema reader.
  */
 public class OracleSchemaReader extends CommonSchemaReader {
   /**
-   *  Oracle type constants, from Oracle JDBC Implementation.
+   * Oracle type constants, from Oracle JDBC Implementation.
    */
   public static final int INTERVAL_YM = -103;
   public static final int INTERVAL_DS = -104;
   public static final int TIMESTAMP_TZ = -101;
   public static final int TIMESTAMP_LTZ = -102;
+  public static final int BINARY_FLOAT = 100;
+  public static final int BINARY_DOUBLE = 101;
+  public static final int BFILE = -13;
+  public static final int LONG = -1;
+  public static final int LONG_RAW = -4;
 
   public static final Set<Integer> ORACLE_TYPES = ImmutableSet.of(
     INTERVAL_DS,
     INTERVAL_YM,
     TIMESTAMP_TZ,
-    TIMESTAMP_LTZ
+    TIMESTAMP_LTZ,
+    BINARY_FLOAT,
+    BINARY_DOUBLE,
+    BFILE,
+    LONG,
+    LONG_RAW,
+    Types.NUMERIC,
+    Types.DECIMAL
   );
 
   @Override
   public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLException {
     int sqlType = metadata.getColumnType(index);
 
-    if (ORACLE_TYPES.contains(sqlType)) {
-      if (sqlType == TIMESTAMP_LTZ || sqlType == TIMESTAMP_TZ) {
+    switch (sqlType) {
+      case TIMESTAMP_TZ:
+      case TIMESTAMP_LTZ:
         return Schema.of(Schema.LogicalType.TIMESTAMP_MICROS);
+      case BINARY_FLOAT:
+        return Schema.of(Schema.Type.FLOAT);
+      case BINARY_DOUBLE:
+        return Schema.of(Schema.Type.DOUBLE);
+      case BFILE:
+      case LONG_RAW:
+        return Schema.of(Schema.Type.BYTES);
+      case INTERVAL_DS:
+      case INTERVAL_YM:
+      case LONG:
+        return Schema.of(Schema.Type.STRING);
+      case Types.NUMERIC:
+      case Types.DECIMAL:
+        // This is the only way to differentiate FLOAT/REAL columns from other numeric columns, that based on NUMBER.
+        // Since in Oracle FLOAT is a subtype of the NUMBER data type, 'getColumnType' and 'getColumnTypeName' can not
+        // be used.
+        if (Double.class.getTypeName().equals(metadata.getColumnClassName(index))) {
+          return Schema.of(Schema.Type.DOUBLE);
+        } else {
+          return super.getSchema(metadata, index);
+        }
+      default:
+        return super.getSchema(metadata, index);
+    }
+  }
+
+  @Override
+  public boolean isTypeCompatible(Schema.Field field, ResultSetMetaData metadata, int index) throws SQLException {
+
+    Schema outputFieldSchema = getSchema(metadata, index);
+    Schema outputFieldNonNullableSchema = outputFieldSchema.isNullable()
+      ? outputFieldSchema.getNonNullable()
+      : outputFieldSchema;
+    Schema inputFieldNonNullableSchema = field.getSchema().isNullable()
+      ? field.getSchema().getNonNullable()
+      : field.getSchema();
+    // Handle the case when output schema expects decimal logical type but we got valid primitive.
```

Comment:
When does this happen?

Reply:
In the case when we specify explicit output schema. 

Since all Oracle numeric types are based on NUMBER(i.e. INTEGER type is actually NUMBER(38, 0)) we won't be able to use Oracle Sink Plugin with other sources.
It's safe to write primitives as values of decimal logical type in the case of valid precision.

Thus the following schema compatibility is supported:
1) Schema.Type.INT -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 10)
2) Schema.Type.LONG -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 19)
3) Schema.Type.FLOAT -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)
4) Schema.Type.DOUBLE -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)

Reply:
I see. I guess this is where you ran into that UI bug where there is no way to specify the precision and scale?

Can you add this information to the comment in the code?

Reply:
Yes, I faced the UI bug testing the Oracle Sink with other sources. 

[Added the info to the comment](https://github.com/data-integrations/database-plugins/pull/46/files#diff-02bcda70486e82bad237c444b0d7f592R66):
```
    // Handle the case when explicit output schema expects decimal logical type but we got valid primitive.
    // Since all Oracle numeric types are based on NUMBER(i.e. INTEGER type is actually NUMBER(38, 0)) we won't be able
    // to use Oracle Sink Plugin with other sources. It's safe to write primitives as values of decimal logical type in
    // the case of valid precision.
    // Thus the following schema compatibility is supported:
    // 1) Schema.Type.INT -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 10)
    // 2) Schema.Type.LONG -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 19)
    // 3) Schema.Type.FLOAT -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)
    // 4) Schema.Type.DOUBLE -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)
    if (Schema.LogicalType.DECIMAL == outputFieldNonNullableSchema.getLogicalType()) {
      int precision = metadata.getPrecision(index);
      switch (inputFieldNonNullableSchema.getType()) {
    ...
```